File: LibretroRT.FrontendComponents.MonoGameRenderer/RenderTargetManager.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using System.Collections.Generic;

namespace LibretroRT.FrontendComponents.MonoGameRenderer
{
    internal class RenderTargetManager : IDisposable
    {
        private const uint CoreRenderTargetMinSize = 1024;

        private static readonly IReadOnlyDictionary<PixelFormats, SurfaceFormat> PixelFormatsMapping = new Dictionary<PixelFormats, SurfaceFormat>
        {
            { PixelFormats.FormatXRGB8888, SurfaceFormat.Color },
            { PixelFormats.FormatRGB565, SurfaceFormat.Bgr565 },
            { PixelFormats.Format0RGB1555, SurfaceFormat.Bgra5551 },
        };

        private static readonly IReadOnlyDictionary<SurfaceFormat, int> PixelFormatsSizeMapping = new Dictionary<SurfaceFormat, int>
        {
            { SurfaceFormat.Color, 4 },
            { SurfaceFormat.Bgr565, 2 },
            { SurfaceFormat.Bgra5551, 2 },
        };

        private readonly object RenderTargetLock = new object();
        private Texture2D RenderTarget = null;
        private Point RenderTargetViewportSize = new Point();
        private int RenderTargetPixelSize = 0;

        public void Dispose()
        {
            RenderTarget.Dispose();
        }

        public void Render(SpriteBatch spriteBatch, Point viewportSize)
        {
            if (RenderTarget == null || RenderTargetViewportSize.X <= 0 || RenderTargetViewportSize.Y <= 0)
                return;

            lock (RenderTargetLock)
            {
                var destinationRect = ComputeBestFittingPosition(viewportSize, RenderTargetViewportSize);
                spriteBatch.Draw(RenderTarget, destinationRect, new Rectangle(Point.Zero, RenderTargetViewportSize), Color.White);
            }
        }

        public void UpdateFromCoreOutput(byte[] frameBuffer, uint width, uint height, uint pitch)
        {
            if (frameBuffer == null || RenderTarget == null)
                return;

            lock (RenderTargetLock)
            {
                var virtualWidth = (int)(pitch / RenderTargetPixelSize);
                RenderTargetViewportSize.X = (int)width;
                RenderTargetViewportSize.Y = (int)height;

                var targetArea = new Rectangle(0, 0, virtualWidth, RenderTargetViewportSize.Y);
                RenderTarget.SetData(0, targetArea, frameBuffer, 0, frameBuffer.Length);
            }
        }

        public void UpdateFormat(GraphicsDevice resourceCreator, GameGeometry geometry, PixelFormats format)
        {
            var requestedFormat = PixelFormatsMapping[format];
            RenderTargetPixelSize = PixelFormatsSizeMapping[requestedFormat];

            if (RenderTarget != null)
            {
                var currentSize = RenderTarget.Bounds.Size;
                if (currentSize.X >= geometry.MaxWidth && currentSize.Y >= geometry.MaxHeight && RenderTarget.Format == requestedFormat)
                {
                    return;
                }
            }

            lock (RenderTargetLock)
            {
                var size = (int)Math.Max(Math.Max(geometry.MaxWidth, geometry.MaxHeight), CoreRenderTargetMinSize);
                size = ClosestGreaterPowerTwo(size);
                var buffer = new byte[size * size * RenderTargetPixelSize];

                RenderTarget?.Dispose();
                RenderTarget = new Texture2D(resourceCreator, size, size, false, requestedFormat);
            }
        }

        private static Rectangle ComputeBestFittingPosition(Point containerSize, Point contentSize)
        {
            var contentAspectRatio = contentSize.X / (float)contentSize.Y;

            Rectangle output;
            int candidateWidth = (int)(containerSize.Y * contentAspectRatio);
            if (containerSize.X >= candidateWidth)
            {
                var size = new Point(candidateWidth, containerSize.Y);
                output = new Rectangle(new Point((containerSize.X - candidateWidth) / 2, 0), size);
            }
            else
            {
                var height = (int)(containerSize.X / contentAspectRatio);
                var size = new Point(containerSize.X, height);
                output = new Rectangle(new Point(0, (containerSize.Y - height) / 2), size);
            }

            return output;
        }

        private static int ClosestGreaterPowerTwo(int value)
        {
            int output = 1;
            while (output < value)
            {
                output *= 2;
            }

            return output;
        }
    }
}

File: LibretroRT.FrontendComponents.Win2DRendererNative/RenderTargetManager.cpp
#include "pch.h"
#include "RenderTargetManager.h"

using namespace LibretroRT_FrontendComponents_Win2DRendererNative;

RenderTargetManager::RenderTargetManager(CanvasAnimatedControl^ canvas) :
	Canvas(canvas)
{
	auto device = GetWrappedResource<ID2D1Device1>(canvas->Device);
	
}


RenderTargetManager::~RenderTargetManager()
{
}

void RenderTargetManager::SetGameGeometry(GameGeometry^ geometry)
{
}

void RenderTargetManager::SetPixelFormat(PixelFormats pixelFormat)
{
}

void RenderTargetManager::UpdateFromCoreOutput(const Array<byte>^ frameBuffer, unsigned int width, unsigned int height, unsigned int pitch)
{
}

void RenderTargetManager::Render(CanvasDrawingSession^ drawingSession, Size canvasSize)
{
}

Rect RenderTargetManager::ComputeBestFittingSize(Size viewportSize, float aspectRatio)
{
	auto candidateWidth = std::floor(viewportSize.Height * aspectRatio);
	if (viewportSize.Width >= candidateWidth)
	{
		Size size(candidateWidth, viewportSize.Height);
		Rect output(Point((viewportSize.Width - candidateWidth) / 2, 0), size);
		return output;
	}
	else
	{
		auto height = viewportSize.Width / aspectRatio;
		Size size(viewportSize.Width, height);
		Rect output(Point(0, (viewportSize.Height - height) / 2), size);
		return output;
	}
}

unsigned int RenderTargetManager::ClosestGreaterPowerTwo(unsigned int value)
{
	unsigned int output = 1;
	while (output < value)
	{
		output *= 2;
	}

	return output;
}
